### ISSUE TITLE
Export process runs indefinitely with "includeAllCustom": true

### ISSUE THREAD

**Issue by homerlex** [eltoroit/ETCopyData#2]
Last night I tried running an export with `"includeAllCustom": true`

This morning when I came in the process was still running and there were no output files created.

I ran the command with loglevel=trace and the last message printed out was about schema discovery:

```
13:43:38.670Z   146     TRACE   SchemaDiscovery.js:229  [PTDataDestination1] Found sObject [MyCustomObject__c].
```

Our schema is fairly complex (multiple levels deep).  Any thoughts on debugging this (or is there a known issue)?  My next step will be to start manually adding custom objects to  `sObjectsData` and see how far I can get without it "hanging."

**Comment by eltoroit** [comment]
I have heard from one more person that the process is hanging. I do not understand why. Are you using a scratch org? If so, could you send me the SFDX project, or at least your sBojects folder (this is safe because you are not sending me the  data, just the metadata). I will like to troubleshoot this issue.

Are you using a windows or Mac?

**Comment by homerlex** [comment]
I'm running on Windows 10.

Yes, I am running the import against a scratch org.  The org was created via sfdx and managed packages (multiple packages) were installed.  I don't have an SFDX project setup.  Unfortunately it won't be possible to get you setup with what we have.  I'll see if I can duplicate with a simpler scenario.

If you want to provide a custom version of your code with more debugs in it I'd bee willing to run that for you.

FYI - I saw the chatter post about the other person having the same issue with his Opportunities.  I did not have any trouble exporting Opps myself.

**Comment by AikeaGuinea** [comment]
I have the same issue when including Quotes or specifying `"includeAllCustom": true`. The command hangs, one core is at 100% and never returns

I am on Linux, cloned the repo and installed the plugin in DX

I include the config file, the output of the command and a dump of the directory of the source (files from the previous run) [case.zip](https://github.com/eltoroit/ETCopyData/files/2699782/case.zip)

I tried to run in debug mode using ` /home/u409105/.local/share/sfdx/client/bin/node --inspect /home/u409105/.local/share/sfdx/client/bin/sfdx.js ETCopyData:export -s dev05 -d dev04 --loglevel trace` but I am unable to attach Chrome to my node (really sorry, no experience with Node :-/)

Is there something I can try?

**Comment by eltoroit** [comment]
Thanks for helping me debug this @AikeaGuinea. Your analysis makes a lot of sense...

How to solve this? If you were not using a tool like this one, how would you load your quotes and your opportunities in a sandbox? Maybe I could code that into the tool...

**Comment by AikeaGuinea** [comment]
Good question! Actually I'd like to know how Salesforce himself does this in the `force:data:tree` (don't know if it is even possible), or through the data API's. I will also ask my colleagues who are far more familiar with Salesforce than myself (be patient focus is elsewhere! ).

But at this stage we are not even trying to import anything, we are just trying to figure out the future import. And I also have issues with e.g. `PricebookEntry` which is flagged as non deletable and non searchable in the `SchemaDiscovery` class and so is non exportable

I think I'll fork and first try to export an environment and import in a empty scratchorg, so that we don't have to worry about the existing, nor have to delete anything

**Comment by AikeaGuinea** [comment]
@eltoroit: After a small break (happy new year btw!) I looked again at our issue. I forked your code and in my `develop` branch I now use an external library to resolve the object graph. It is still in early stage.

About the circular references: I discussed this with an admin and actually they would be happy to import the Opportunity and the Quote and can live without the synchedQuote back reference. In our case we are trying to populate scratchorgs  with default data, we don't make full copy of data between sandboxes.

For the time being, it is best to fully avoid this circular behaviour (currently: error thrown). The object graph resolver can handle this (but I had issues). Ideally when including objects that rely on each other (Quote and Opportunity), the config should exclude one of the references (e.g. by specifying ` IgnoreFields: SynchedQuote.Id`). We should then detect this, which is not really easy, and omit the dependency. That will be my next attempt.

**Comment by homerlex** [comment]
I found that we also have schema objects with circular references and this is why we hang  (infinite loop I suspect) when `"includeAllCustom": true`

**Comment by homerlex** [comment]
What is the status of the circular reference issue?

**Comment by eltoroit** [comment]
The issue has been fixed with commit #30 thanks to @OscarScholten.
